runtime/lua/vim/_system.lua

```diff
@@ -0,0 +1,373 @@
+end
+
+local MAX_TIMEOUT = 2 ^ 31
+
+--- @param timeout? integer
+--- @return SystemCompleted
+function SystemObj:wait(timeout)
+  local state = self._state
+
+  vim.wait(timeout or state.timeout or MAX_TIMEOUT, function()
+    return state.done
+  end)
+
+  if not state.done then
+    self:kill(6) -- 'sigint'
+    state.result = timeout_result(state.cmd, state.args)
+  end
+
+  return state.result
+end
+
+--- @param data string[]|string|nil
+function SystemObj:write(data)
+  local stdin = self._state.stdin
+
+  if not stdin then
+    error('stdin has not been opened on this object')
+  end
+
+  if type(data) == 'table' then
+    for _, v in ipairs(data) do
+      stdin:write(v)
+      stdin:write('\n')
+    end
+  elseif type(data) == 'string' then
+    stdin:write(data)
+  elseif data == nil then
+    -- Shutdown the write side of the duplex stream and then close the pipe.
+    -- Note shutdown will wait for all the pending write requests to complete
+    -- TODO(lewis6991): apparently shutdown doesn't behave this way.
+    -- (https://github.com/neovim/neovim/pull/17620#discussion_r820775616)
+    stdin:write('', function()
+      stdin:shutdown(function()
+        if stdin then
+          stdin:close()
+        end
+      end)
+    end)
+  end
+end
+
+--- @return boolean
+function SystemObj:is_closing()
+  local handle = self._state.handle
+  return handle == nil or handle:is_closing()
+end
+
+--- @param spec string[]|SystemSpec
+--- @return SystemSpec
+local function process_spec(spec)
+  if spec[1] then
+    local cmd = {} --- @type string[]
+    for _, p in ipairs(spec) do
+      cmd[#cmd + 1] = p
+    end
+    spec = vim.deepcopy(spec)
+    spec.cmd = cmd
+  end
+
+  return spec
+end
+
+---@private
+---@param output function|'false'
+---@return uv_stream_t?
+---@return function? Handler
+local function setup_output(output)
+  if output == nil then
+    return assert(uv.new_pipe(false)), nil
+  end
+
+  if type(output) == 'function' then
+    return assert(uv.new_pipe(false)), output
+  end
+
+  assert(output == false)
+  return nil, nil
+end
+
+---@private
+---@param input string|string[]|true|nil
+---@return uv_stream_t?
+---@return string|string[]?
+local function setup_input(input)
+  if not input then
+    return
+  end
+
+  local towrite --- @type string|string[]?
+  if type(input) == 'string' or type(input) == 'table' then
+    towrite = input
+  end
+
+  return assert(uv.new_pipe(false)), towrite
+end
+
+--- @param cmd string|string[]
+--- @param args? string[]
+--- @return string Command
+--- @return string[]? Arguments
+local function setup_cmd(cmd, args)
+  if type(cmd) == 'string' then
+    cmd = { cmd }
+    if args then
+      vim.list_extend(cmd, args)
+    end
+  end
+
+  return cmd[1], vim.list_slice(cmd, 2)
+end
+
+--- @return table<string,string>
+local function base_env()
+  local env = vim.fn.environ()
+  env['NVIM_LISTEN_ADDRESS'] = nil
+  env['NVIM_LOG_FILE'] = nil
+  env['VIM'] = nil
+  env['VIMRUNTIME'] = nil
```

Comment:
Could leave these 3 as a TODO if they cause trouble. They are still TODO for jobstart() I think.

Reply:
I've removed them and just documented `NVIM`.